pandas/tests/io/test_sql.py

```diff
@@ -2698,6 +2700,20 @@ def test_drop_table(conn, request):
         assert not insp.has_table("temp_frame")
 
 
+@pytest.mark.parametrize("conn", sqlalchemy_connectable + ["sqlite_buildin"])
+def test_delete_rows_success(conn, test_frame1, request):
+    table_name = "temp_frame"
+    conn = request.getfixturevalue(conn)
+    pandasSQL = pandasSQL_builder(conn)
+    with pandasSQL.run_transaction():
+        assert (
+            pandasSQL.to_sql(test_frame1, table_name, if_exists="replace")
+            == test_frame1.shape[0]
+        )
+        assert pandasSQL.delete_rows(table_name) is None
+        assert count_rows(conn, table_name) == 0
+
+
 @pytest.mark.parametrize("conn", all_connectable)
```

Comment:
Can you add a test that ensures that this is atomic?

Reply:
Sorry, the location of this comment got me confused. You mean testing `delete_rows` is atomic ?

Reply:
Yea; I think can add a new test (unless you see a good one that can be parametrized)

Reply:
I'll need some help here. I failed to get the transaction in which `delete_rows` is running.

Reply:
I don't think you need to inspect the transaction; that's probably a few layers removed from where we are operating.

I think you can just write some data, change some integral columns to string, and try to rewrite the same data. I think this should fail, and afterwards you can inspect the table to make sure the data that was successfully written remains

Reply:
Yeah.... a caveat is this won't work for sqlite connections due to its dynamic typing. I don't think we support strict tables, correct ?
WDYT ?

Reply:
Wouldn't it still fail since the table structure for SQLite is expecting an INT but we send TEXT? If not then we can xfail just for SQLite

Reply:
Yes, I was surprise by this behavior but learning it is a [feature](https://sqlite.org/datatype3.html). It is possible to use the `strict` mode when creating the table neither of pandas or sqlalchemy support that as of now.
![flexible-type](https://github.com/user-attachments/assets/3b96c065-6a9f-4b38-b40c-50b16bd4d511)
![flexible-type-2](https://github.com/user-attachments/assets/dfea88fa-1c52-4747-9715-177d6cfd5455).
---
Other than that I've been scratching my head the last couple of hours. I was able to assert the atomic behavior for the SQLAlchemy backend but failed miserably for the ADBC ones. Still trying to grok what's going on behind scenes. Any hunch ?

Reply:
Thanks for the clarification on SQLite. So if you were to inspect the type of the column after performing a DELETE operation, would it retain the data type prior to the DELETE or what you append after? I suppose there's some cause for concern there, as this DELETE / Append might introduce different behaviors across different databases, which is something we need to avoid.

Not sure about the ADBC portion, but sounds like the transaction management in the implementation is off

Reply:
> Thanks for the clarification on SQLite. So if you were to inspect the type of the column after performing a DELETE operation, would it retain the data type prior to the DELETE or what you append after? I suppose there's some cause for concern there, as this DELETE / Append might introduce different behaviors across different databases, which is something we need to avoid.

Yes, it would retain the original type. This affects all databases with dynamic types (SQLite is the only I know). Are you thinking of adding safeguards for both delete/append to avoid that ? Maybe we should document that behavior instead of blocking something a user might want to have as it is provided by the database itself ?

> Not sure about the ADBC portion, but sounds like the transaction management in the implementation is off
Yes, and this is what is bothering me. I'll push the latest changes would be great if you could take a look ?

Reply:
Hmm if we can't get consistent behavior  I'm not sure its worth pursuing here. This is just a small convenience of the user calling DELETE themselves, but for us to take on and manage expectations around diverging database behavior would be pretty far out of scope for the pandas project

Reply:
Hello again o/,

Tests are working now.

> Hmm if we can't get consistent behavior I'm not sure its worth pursuing here.

Pandas is consistent. Be it using `append` (prone to the same problem) or `delete_rows`. The naming `delete_rows` implies we're not going to change the schema of a table. If the database you're using supports such "non-standard" behavior (appending / deleting different types) that is up to you to know - pandas should not interfere on that matter.

>  This is just a small convenience of the user calling DELETE themselves

I do agree and we're consistent on that matter. We delete all rows of the table in a transaction - no matter the database - and insert the new records.

Reply:
Hello @WillAyd , I hope this message finds you well.
Do you have any final word on this PR ?

Reply:
Thanks for the ping and sorry for delay in response - I've had a busy few weeks personally.

OK I think I'm on board with what you are saying just need to carve out some time to review in more detail. I'll try to do that in the next few days - I appreciate your patience and thanks for working on this

Reply:
No worries.

> OK I think I'm on board with what you are saying just need to carve out some time to review in more detail.

Oh, that's nice. I'm happy to hear that.

> I'll try to do that in the next few days - I appreciate your patience and thanks for working on this

Thanks a lot ! :)